design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,163 @@
+
+An alternative is to let the user dictate the prefixes. This way, they can optimize them knowing beforehand what virtual chunks they will add. If they don't optimize, they get large manifests without prefix extraction. We still need to satisfy requirements 2 and 4, which would require automatic extraction of at least the bucket and platform.
+
+Additionally, the optimization problem can be treated as an extra process (similar to compaction or garbage collection), and ran offline on demand.
+
+### Templates instead of prefixes
+
+Using templates would be more efficient, example:
+
+* `s3://some-bucket/some-prefix/1-analysis-output-for-model-abc.nc`
+* `s3://some-bucket/some-prefix/2-analysis-output-for-model-abc.nc`
+
+If we use the template `s3://some-bucket/some-prefix/{}-analysis-output-for-model-abc.nc` each chunk could have a couple of extra characters only, filling the placeholders in the template.
+
+### Supporting multiple clouds
+
+When the user adds a virtual ref, they need to indicate the details of the platform to retrieve it. These include what cloud it is (so we know how to set up credentials), region, endpoint url, port and other settings. It looks a lot like our `S3Config` type (or an expanded version of that).
+
+### Supporting multiple credentials
+
+Credentials are a read time concern. When the user writes the virtual chunk, they don't necessarily know what credentials set will be used to read it.
+
+Credentials need to be specified later, when the chunks are going to be read. For example, by indicating in runtime configuration a set of credentials for each virtual chunk bucket/prefix/template.
+
+## Design
+
+### Explicit containers for virtual chunks
+
+* Each repo has a global list of `VirtualChunkContainers`
+
+```rust
+// this is pseudo-code
+struct VirtualChunkContainer {
+  name: String,  // something like model-output
+  platform: Platform,  // something like S3
+  url_template: String, // something like some-bucket/some-prefix/output/{}/{}.nc
+  default_template_arguments: Vec<String>,
+  cloud_client_config: Map<String, Any>,  // something like {endpoint_url: 8080}
+}
+```
+
+* Users can add to the list and edit members but not delete.
+
+### Global config
+
+* The list of chunk containers is stored in a new global `Config` object. There is a single configuration per repo, and all reads and writes are done reading the latest configuration. In the initial implementation, the "history" of configuration is present only for reference and user inspection by hand.
+
+TBD: how to store this information. One approach would be using the new PUT if-match S3 functionality, but this would be the first place where we overwrite objects.
+
+```rust
+// this is pseudo-code
+struct Config {
+  inline_chunk_threshold_bytes: u16,
+  virtual_chunk_containers: Vec<VirtualChunkContainer>,
+}
```

Comment:
In general, I really like the idea of repo-level config stored with the repo. Iceberg has this and it seems really useful. Could eventually put all kinds of stuff here.